log/loki.go

```diff
@@ -0,0 +1,499 @@
+			}
+		case "level":
+			h.levels, err = getLevels(value)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			if strings.HasPrefix(key, "label.") {
+				labelKey := strings.TrimPrefix(key, "label.")
+				h.labels = append(h.labels, [2]string{labelKey, value})
+				continue
+			}
+
+			return nil, fmt.Errorf("unknown loki config key %s", key)
+		}
+	}
+
+	h.client = &http.Client{Timeout: h.pushPeriod}
+
+	h.start()
+	return h, nil
+}
+
+func getLevels(level string) ([]logrus.Level, error) {
+	// TODO figure out if `tracing`,`fatal` and `panic` should be included
+	levels := []logrus.Level{}
+	switch level {
+	case "debug":
+		levels = append(levels, logrus.DebugLevel)
+		fallthrough
+	case "info":
+		levels = append(levels, logrus.InfoLevel)
+		fallthrough
+	case "warning":
+		levels = append(levels, logrus.WarnLevel)
+		fallthrough
+	case "error":
+		levels = append(levels, logrus.ErrorLevel)
+	default:
+		return nil, fmt.Errorf("unknown log level %s", level)
+	}
+	return levels, nil
+}
+
+func (h *lokiHook) start() {
+	h.ch = make(chan *logrus.Entry, 1000)
+	go h.loop()
+}
+
+// fill one of two equally sized slices with entries and then push it while filling the other one
+// TODO benchmark this
+//nolint:funlen
+func (h *lokiHook) loop() {
+	var (
+		msgs       = make([]tmpMsg, h.limit)
+		msgsToPush = make([]tmpMsg, h.limit)
+		dropped    int
+		count      int
+		ticker     = time.NewTicker(h.pushPeriod)
+		pushCh     = make(chan chan time.Duration)
+	)
+
+	defer ticker.Stop()
+	defer close(pushCh)
+
+	go func() {
+		oldLogs := make([]tmpMsg, 0, h.limit*2)
+		for ch := range pushCh {
+			msgsToPush, msgs = msgs, msgsToPush
+			oldCount, oldDropped := count, dropped
+			count, dropped = 0, 0
+			bufferTime := <-ch
+			close(ch) // signal that more buffering can continue
+
+			copy(oldLogs[len(oldLogs):len(oldLogs)+oldCount], msgsToPush[:oldCount])
+			oldLogs = oldLogs[:len(oldLogs)+oldCount]
+
+			t := time.Now()
+			cutOffIndex := sortAndSplitMsgs(oldLogs, bufferTime)
+			if cutOffIndex == 0 {
+				continue
+			}
+			t1 := time.Since(t)
+
+			strms := h.createPushMessage(oldLogs, cutOffIndex, oldDropped)
+			if cutOffIndex > len(oldLogs) {
+				oldLogs = oldLogs[:0]
+				continue
+			}
+			oldLogs = oldLogs[:copy(oldLogs, oldLogs[cutOffIndex:])]
+			t2 := time.Since(t) - t1
+
+			var b bytes.Buffer
+			_, err := strms.WriteTo(&b)
+			if err != nil {
+				fmt.Printf("Error while marshaling logs for loki %s\n", err)
+				continue
+			}
+			size := b.Len()
+			t3 := time.Since(t) - t2 - t1
+
+			err = h.push(b)
+			if err != nil {
+				fmt.Printf("Error while sending logs to loki %s\n", err)
+				continue
+			}
+			t4 := time.Since(t) - t3 - t2 - t1
+
+			if h.profile {
+				fmt.Printf("sorting=%s, adding=%s marshalling=%s sending=%s count=%d final_size=%d\n",
+					t1, t2, t3, t4, cutOffIndex, size)
+			}
+		}
+	}()
+
+	for {
+		select {
+		case entry := <-h.ch:
+			if count == h.limit {
```

Comment:
I think the check here should be `if h.limit > 0 && count == h.limit`, to more explicitly allow for unlimited messages when `limit` is `0`, though I guess you can do it even now with `-1`...

Reply:
the way the code is currently written I am pretty sure h.limit<1 will panic ;) Maybe something that should be an error during the parsing of the config

Reply:
tested:
1 seems to work
0 means nothing is send(including the dropped message)
-1 (and below I would imagine) panics

Reply:
> the way the code is currently written I am pretty sure h.limit<1 will panic

That should probably be considered a bug, so :+1: for some validation. And :+1: for supporting `0` (i.e. no limit), since it seems reasonable to me that if I'm running my own loki instance, I'd like to get all of my log messages. 

As I said before, I understand why the current limit per second exists, but I don't like it one bit. I can't think of any other piece of software that omits log messages from the middle... And while I can live with it as a form of abuse prevention / DoS protection in the cloud, for people self-hosting k6 and loki it doesn't make any sense to have a limit that can eat your logs. Especially because, when loki is enabled, we set the regular output of the log to `io.Discard` (which I'm not sure is the correct action anymore...).

Reply:
I will leave the "unlimited" for another PR - the current implementation is fast(er) because it uses the fact that I have an upper limit to the messages. Arguably anyone can set the limit to 1000000 and ... not hit it, even if they do all the other things in the current implementation will mean that that push to loki will take forever.

Reply:
Hmm if you've set the limit to `1000000`, you'd needlessly allocate a couple of slices with 1000000 elements, right? This seems like a premature optimization that might not always work now used as a justification for always requiring the limit :man_shrugging: But with 1000000 or even 10000, you'd likely never reach that many messages per second, so it'd probably have the opposite effect...

I like that the `msgs` buffers are reused, but wouldn't it be better to not have a limit and just `append()` stuff to the buffer, allowing it to grow naturally to its optimal size? After the first few extensions, it's likely to reach the limit or some equilibrium, if there's no limit?

Btw, not that it matters all that much, but with the current strategy of dropping log messages from the middle of the stream, and with messages arriving somewhat out of order, it will lead to the following strange situation. Say the limit is 100 messages per second, and the user is sending ~150. Normally, they'd see the first 100 messages every second, followed by a `k6 dropped some packages because they were above the limit of 100/1s` message for the remaining ~50. But, because the sorting happens after the dropping, some of these dropped messages could have arrived out-of-order and technically be with timestamps from before the `k6 dropped some...` message :disappointed:

Reply:
well yes ... but now imagine that the user said they thought they will get  10000 messages and they got 10000 times more ... they just ran out of memory ...  the process died, they not only didn't get 10000 messages they got 0 messages and a big stacktrace.

Organically growing and unlimited are things that just lead to the software running out of memory and dying of that. 

In a normal (perfectly running) and well-configured k6 none of this matters - we might have allocated 10x/100x/1000x times more space for eventual logs and handling them. But at least in the (more likely IMO scenario) where something isn't perfect and it generates 1000x more logs we don't suddenly have an ever-expanding list of logs ... which just kills k6 for no good reason and no explanation

Reply:
I didn't request that you remove the `limit` altogether, or grow the `msgs` buffer infinitely if there's a `limit`, I even said "I understand why the current limit per second exists", so don't become snarky please :stuck_out_tongue: 

To be even more clear, I'm not suggesting that you remove this code completely: https://github.com/loadimpact/k6/blob/5f39dc2ef5c070b74c6b7d0fd7959ee323af071d/log/loki.go#L213-L216

Just replace it with `if h.limit > 0 && count == h.limit`, ditch the fixed buffer sizes, and use `append(msgs, tmpMsg{ ... })` instead of `msgs[count] = tmpMsg{...}`. 

This way, if you have a `limit` (which would still be the case by default), the `msgs` buffer will never be able to grow above it, right? Or whatever power of 2 is above it, I guess, given how Go grows slices. But, if someone deliberately chooses to remove the limit completely, yes, they can potentially shoot themselves in the foot. Though, to be fair, there are probably a ton of other easier ways to run out of memory with k6 :wink:

Reply:
Just read through this discussion + the one on Slack, and you both make valid points, so not sure who I agree more with, but here are my €0.02. :sweat_smile:

The current `limit` is essentially a throttling feature, right? It tries to accomplish two things: don't overload Loki, and don't run out of memory in k6.

I agree with Ned's concerns about the current implementation dropping logs if the limit is reached, which you definitely *never* want (if it can be avoided), especially if it's statically defined for the entire process runtime. Consider that Loki's performance might vary over time, so this package artifically limiting the throughput even if Loki itself could handle more messages, or viceversa, sending more messages than Loki can handle because the limit was set too high and Loki is temporarily struggling with other clients... is not good.

If we do want to introduce something like this, then I would argue it should be more sophisticated:

1. React either on current RAM usage (probably has to wait for #888) and/or some type of Loki healthcheck or heuristic we can use to determine it's becoming "overloaded" (e.g. response times), and factor that into calculating what the next payload size should be and when to push the next request. Sort of like a smart backoff.
2. Don't just drop logs, but save them in a third "missed" buffer, and gradually feed them in subsequent requests. I think this should be fixed even if we don't implement the more advanced throttling.

I'm torn on being able to specify "unlimited" as an option. On the one hand we need some type of hard limit in the Cloud to prevent one test from flooding the Loki instance (BTW, how does this work in the backend exactly? It's one central instance for all tests?), but we don't care for it during development. I wouldn't mind having to set e.g. `limit=1000000` to make that work, though.

Reply:
Even after a week of vacation, I still think the most sensible choice is for the default `limit` to be `0` (i.e. "unlimited") and that the current fixed-length buffer approach is a case of way premature optimization :wink: It even makes more sense when you reverse the question, instead of asking "why not 0", to ask "why 100 and not 57 or 114?"... or "why not 50 or 64", if we're fans of round numbers :wink: 

That said, it's a minor issue in a feature most people probably won't use, so I don't really care if you're so determined to keep the current defaults. This is not a blocking issue from my side, in contrast to https://github.com/loadimpact/k6/pull/1576/files#r463614996